src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+                     << cpp_strerror(r) << dendl;
+        return r;
+      }
+    }
+  }
+
+  ldout(m_cct, 20) << "removing migration header" << dendl;
+
+  r = cls_client::migration_remove(&m_image_ctx->md_ctx,
+                                   m_image_ctx->header_oid);
+  if (r == -ENOENT) {
+    r = 0;
+  }
+  if (r < 0) {
+    lderr(m_cct) << "failed removing migration header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::v2_unlink_image() {
+  ldout(m_cct, 10) << dendl;
+
+  m_image_ctx->owner_lock.get_read();
+  if (m_image_ctx->exclusive_lock != nullptr &&
+      m_image_ctx->exclusive_lock->is_lock_owner()) {
+    C_SaferCond ctx;
+    m_image_ctx->exclusive_lock->release_lock(&ctx);
+    m_image_ctx->owner_lock.put_read();
+    int r = ctx.wait();
+    if (r < 0) {
+      lderr(m_cct) << "error releasing exclusive lock: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+  } else {
+    m_image_ctx->owner_lock.put_read();
+  }
+
+  int r = trash_move(m_src_io_ctx, RBD_TRASH_IMAGE_SOURCE_MIGRATION,
+                     m_image_ctx->name, 0);
+  if (r < 0) {
+    lderr(m_cct) << "failed moving image to trash: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::v2_relink_image() {
+  int r = trash_restore(m_src_io_ctx, m_image_ctx->id, m_image_ctx->name);
+  if (r < 0) {
+    lderr(m_cct) << "failed restoring image from trash: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::create_new_image() {
+  ldout(m_cct, 10) << dendl;
+
+  uint64_t size;
+  {
+    RWLock::RLocker snap_locker(m_image_ctx->snap_lock);
+    size = m_image_ctx->size;
+  }
+
+  ThreadPool *thread_pool;
+  ContextWQ *op_work_queue;
+  ImageCtx::get_thread_pool_instance(m_cct, &thread_pool, &op_work_queue);
+
+  C_SaferCond on_create;
+  auto *req = image::CreateRequest<I>::create(
+      m_dst_io_ctx, m_dst_image_name, m_dst_image_id, size, m_image_options, "",
+      "", false, op_work_queue, &on_create);
+  req->send();
+  int r = on_create.wait();
+  if (r < 0) {
+    lderr(m_cct) << "header creation failed: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  auto dst_image_ctx = new librbd::ImageCtx(m_dst_image_name, m_dst_image_id,
+                                            nullptr, m_dst_io_ctx, false);
+  dst_image_ctx->ignore_migrating = true; // XXXMG: hackish
+  r = dst_image_ctx->state->open(false);
+  if (r < 0) {
+    lderr(m_cct) << "failed to open newly created header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  dst_image_ctx->owner_lock.get_read();
+  if (dst_image_ctx->exclusive_lock != nullptr &&
+      !dst_image_ctx->exclusive_lock->is_lock_owner()) {
+    C_SaferCond ctx;
+    dst_image_ctx->exclusive_lock->acquire_lock(&ctx);
```

Comment:
Probably should block API requests while a migration is in-progress (at least until all state is set up).